File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/C#/Samples/General Controller/WaveGeneratorSample/WaveGeneratorSample.cs
/** <copyright file="WavegeneratorSample.cs" company="Physik Instrumente (PI) SE & Co. KG">
 *  (c)2015-2020 Physik Instrumente (PI) SE & Co. KG
 *  Software products that are provided by PI are subject to the General Software License Agreement of
 *  Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components.
 *  For more information, please read the General Software License Agreement and the Third Party Software Note linked below.
 *
 *  General Software License Agreement :
 *  http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
 *  Third Party Software Note :
 *  http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
 *  </copyright>*/


/**<remarks>
 *   !! MOTION WARNING !!
 *
 *   This sample connects to a controller using the PI Connection Dialog. The connection type can be easily changed.
 *   Then the wave generator is configured and started.
 *   The stage has to be initialized and referenced before starting this sample.
 *   </remarks> */


using System;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using PI;

namespace WaveGeneratorSample
{
    public enum ConnectionType
    {
        Dialog,
        Rs232,
        Tcpip,
        Usb
    }

    internal class GcsCommandError : Exception
    {
        public GcsCommandError(string message)
            : base(message)
        {
        }
    }

    /** <summary>
     *  This class runs the wave generator for a controller that supports this feature, i.e. E-727, C-887.
     *  Call the Run-method to execute this sample.
     *  Please choose the right configuration for your setup, there is one common configuration
     *  working with most of PI controllers which support the wave generator feature
     *  and one configuration for the Hexapod controller C-887
     * </summary>*/
    class WaveGeneratorSample
    {
        /// <remarks>
        /// select the wave generator configuration here: 'Common' or 'C-887'
        /// </remarks>
        #region ControllerSpecificSettings
        private readonly WaveGeneratorSettings _waveGeneratorSettings = WaveGeneratorConfigurations.Properties["Common"];
        #endregion

        #region fields
        private const int PI_TRUE = 1;
        private const int PI_FALSE = 0;
        private const int PI_RESULT_FAILURE = 0;
        private const uint PI_PARAMETER_ID_SERVO_TIME = 0xE000200;
        internal int ControllerId = -1;
        #endregion

        #region public methods
        public void Run()
        {
            try
            {
                OpenConnection(ConnectionType.Dialog);
                PrintControllerIdentification();

                HasWaveGenerator();
                ConfigureWaveGenerator();

                var state = new[] {PI_TRUE};
                SetServoState(_waveGeneratorSettings.Axis, state);
                GoToStartPosition();
                StartWaveGenerator();
                state = new[] {PI_FALSE};
                SetServoState(_waveGeneratorSettings.Axis, state);

            }
            catch (GcsCommandError e)
            {
                PrintErrorMessage(e);
            }
            catch (Exception e)
            {
                // Catch all Exception which ar not GCSErrors, i.e. for file handling
                Console.WriteLine(e);
            }
            finally
            {
                DisconnectController();
                Environment.ExitCode = -1;
            }
        }
        #endregion

        #region private methods
        private void StartWaveGenerator()
        {
            // restart recording as soon as wave generator output starts running
            if ( GCS2.WGR(ControllerId) == PI_RESULT_FAILURE )
            {
                throw new GcsCommandError("Failed to restart recording.");
            }

            // start wave generator output, data recorder starts simultaneously
            if ( GCS2.WGO( ControllerId
                            , _waveGeneratorSettings.WaveGeneratorId
                            , _waveGeneratorSettings.StartMode
                            , 1)
                 == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to start wave generator output.");
            }

            WaitForWaveGeneratorDone().Wait(7000);
        }

        private async Task WaitForWaveGeneratorDone()
        {
            // initialize provisional array for WG WaveGeneratorId,
            // since PI_IsGeneratorRunning expects an array as parameter
            // status of wave generator (0 = not running, 1 = running)
            var isWgRunning = Enumerable.Repeat( 1 , _waveGeneratorSettings.NumberOfAxis).ToArray();
            var waveGeneratorIdArray = _waveGeneratorSettings.WaveGeneratorId;

            // query wave generator status
            while (isWgRunning[0] == 1)
            {
                await Task.Delay(100);

                if ( GCS2.IsGeneratorRunning( ControllerId
                                            , waveGeneratorIdArray
                                            , isWgRunning
                                            , 1)
                     == PI_RESULT_FAILURE )
                {
                    throw new GcsCommandError("Unable to query wave generator status.");
                }
            }
        }

        private void GoToStartPosition()
        {
            Console.WriteLine("Starting wave generator...");

            // move stage to offset of wave prior to starting wave generator
            if (GCS2.MOV( ControllerId
                    , _waveGeneratorSettings.Axis
                    , _waveGeneratorSettings.OffsetOfWave)
                == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to move stage to start position.");
            }

            WaitForMotionDone(_waveGeneratorSettings.Axis).Wait(7000);
        }

        private async Task WaitForMotionDone (string axis)
        {
            int[] isMoving = {0};

            while ( Array.Exists(isMoving, element => element != 0 ))
            {
                if (GCS2.IsMoving(ControllerId, axis, isMoving ) == PI_RESULT_FAILURE )
                {
                    throw new GcsCommandError("Unable to query movement status.");
                }

                await Task.Delay(10);
            }

            Console.WriteLine("Movement is finished.");

        }

        private void ConfigureWaveGenerator()
        {
            Console.WriteLine("Setting up wave generator: Frequency: "
                              + _waveGeneratorSettings.Frequency
                              + " Hz, Amplitude: "
                              + _waveGeneratorSettings.AmplitudeOfWave
                              + " mm");
            // define waveform

            var servoCycleTime = GetServoCycleTime();

            // calculate number of point in wave table from given frequency
            var numberOfPoints = (int) (1 / (servoCycleTime[0] * _waveGeneratorSettings.Frequency));

            // wave table contains one segment
            var segmentLength = numberOfPoints;

            // curve center point is the middle of the segment
            var centerPointOfWave = numberOfPoints / 2;

            // send wave table to controller

            if (_waveGeneratorSettings.WaveGeneratorId.Any(id => GCS2.WAV_SIN_P(ControllerId
                                                        , id // 1
                                                        , _waveGeneratorSettings.OffsetOfFirstPointInWaveTable // 0
                                                        , numberOfPoints // 2000
                                                        , _waveGeneratorSettings.AppendWave // 0
                                                        , centerPointOfWave // 1000
                                                        , _waveGeneratorSettings.AmplitudeOfWave // 0.01
                                                        , _waveGeneratorSettings.OffsetOfWave[0] //0.0
                                                        , segmentLength //2000
                                                                        )
                                                    == PI_RESULT_FAILURE) )
            {
                throw new GcsCommandError("Unable to set up waveform.");
            }

            // link wave table to wave generator
            if (GCS2.WSL( ControllerId
                    , _waveGeneratorSettings.WaveGeneratorId
                    , _waveGeneratorSettings.WaveTableId
                    , 1)
                == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to link wave table to wave generator.");
            }

            // set wave table rate
            if (GCS2.WTR( ControllerId
                        , new []{0}
                        , _waveGeneratorSettings.TableRate
                        , _waveGeneratorSettings.InterpolationType
                        , _waveGeneratorSettings.TableRate.Length)
                == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to set wave table rate.");
            }

            // set wave generator output cycles
            if (GCS2.WGC( ControllerId
                        , _waveGeneratorSettings.WaveGeneratorId
                        , _waveGeneratorSettings.NumberOfCycles
                        , 1)
                == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to set wave generator output cycles.");
            }
        }

        private double[] GetServoCycleTime()
        {
            // query servo cycle time
            double[] servoCycleTime = {0.0}; // servo cycle time of controller
            uint[] parametersToQuery = {PI_PARAMETER_ID_SERVO_TIME};

            if (GCS2.qSPA( ControllerId
                    , _waveGeneratorSettings.Axis
                    , parametersToQuery
                    , servoCycleTime
                    , null
                    , 0)
                == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable to query servo update time.");
            }

            return servoCycleTime;
        }

        private void HasWaveGenerator()
        {
            if (!CheckNumberOfWaveGenerators())
            {
                throw new GcsCommandError("Controller has no wave Generators");
            }
        }

        private bool CheckNumberOfWaveGenerators()
        {
            var numberOfWaveGenerators = 0;

            if (GCS2.qTWG(ControllerId, ref numberOfWaveGenerators ) == PI_RESULT_FAILURE)
            {
                throw new GcsCommandError("Unable retrieve number of wave generators.");
            }

            return (numberOfWaveGenerators > 0);
        }

        private void OpenConnection(ConnectionType type, string hostName = "localhost")
        {
            Console.WriteLine("-- Try to Connect --");
            switch (type)
            {
                case ConnectionType.Dialog:
                    ControllerId = GCS2.InterfaceSetupDlg("");
                    break;
                case ConnectionType.Rs232:
                    ControllerId = GCS2.ConnectRS232(1, 115200);
                    break;
                case ConnectionType.Tcpip:
                    ControllerId = GCS2.ConnectTCPIP(hostName, 50000);
                    break;
                case ConnectionType.Usb:
                {
                    var connectedUsbController = new StringBuilder(1024);
                    var noDevices = GCS2.EnumerateUSB(connectedUsbController, 1024, "");

                    Console.WriteLine("Found: " + noDevices + " USB controllers, connecting to the first one:");
                    Console.Write(connectedUsbController + "\n");

                    ControllerId = GCS2.ConnectUSB(
                        connectedUsbController
                            .ToString()
                            .Split(new[] {Environment.NewLine}, StringSplitOptions.None)
                            [0]);
                    break;
                }
                default:
                    throw new ArgumentOutOfRangeException(nameof(type), type, null);
            }

            if (0 <= ControllerId) return;

            var connType = type.Equals(ConnectionType.Dialog) ? type.ToString() : type.ToString().ToUpper();
            throw new GcsCommandError("Connection to " + connType + " failed. Exiting.");
        }

        private void DisconnectController()
        {
            if (ControllerId >= 0)
            {
                GCS2.CloseConnection(ControllerId);
                ControllerId = -1;
            }

            Console.WriteLine("Close connection.");
        }

        private void PrintErrorMessage(GcsCommandError e)
        {
            var errorCode = GCS2.GetError(ControllerId);
            var errorMessage = new StringBuilder(1024);
            GCS2.TranslateError(errorCode, errorMessage, errorMessage.Capacity);
            Console.WriteLine("ERROR: " + e.Message + " - Reported error code: " + errorCode + ": " + errorMessage);
        }

        private void PrintControllerIdentification()
        {
            var controllerIdentification = new StringBuilder(1024);

            if (GCS2.qIDN(ControllerId, controllerIdentification, controllerIdentification.Capacity) ==
                PI_RESULT_FAILURE)
            {
                throw (new GcsCommandError("qIDN failed. Exiting."));
            }

            Console.WriteLine("qIDN returned: " + controllerIdentification);
        }

        private void SetServoState(string axis, int[] state)
        {
            if (GCS2.SVO(ControllerId, axis, state ) == PI_RESULT_FAILURE )
            {
                throw (new GcsCommandError("SVO failed. Exiting"));
            }
        }
        #endregion
    }
}

File: Device_control/PI-E709/PI_codes/PI-Software-Suite-C-990.CD1/Development/C++/Samples/General Controller/WaveGeneratorSample/WaveGeneratorSample.cpp
/**
    (c)2015 Physik Instrumente (PI) SE & Co. KG
    Software products that are provided by PI are subject to the General Software License Agreement of 
    Physik Instrumente (PI) SE & Co. KG and may incorporate and/or make use of third-party software components. 
    For more information, please read the General Software License Agreement and the Third Party Software Note linked below.

    General Software License Agreement :
    http://www.physikinstrumente.com/download/EULA_PhysikInstrumenteGmbH_Co_KG.pdf
    Third Party Software Note :
    http://www.physikinstrumente.com/download/TPSWNote_PhysikInstrumenteGmbH_Co_KG.pdf
**/

/**
    !! MOTION WARNING !!

    This sample connects to a controller using a connection dialog.
    Then the wave generator is configured and started.
    The positioner has to be referenced before starting this sample.
    
**/

#include "WaveGeneratorSample.h"


int main ()
{
    int iD = OpenConnection (Dialog);

    if (iD < 0)
    {
        std::cout << "Unable to connect.";
        return 1;
    }

    try
    {
        PrintControllerIdentification (iD);

        // check if controller has a wave generator
        int noWaveGenerators = 0;

        if (!PI_qTWG (iD, &noWaveGenerators))
        {
            throw std::runtime_error ("Unable to retrieve number of wave generators.");
        }

        if (noWaveGenerators > 0)
        {
            std::string sAxis ("1");
            // choose from the predefined parameter sets defined in the header file
            waveGeneratorSettings wgConfig = commonSettings;

            SetServoState (iD, sAxis, SERVO_ON);
            StartWaveGenerator (iD, sAxis, wgConfig);
            SetServoState (iD, sAxis, SERVO_OFF);
        }
        else
        {
            std::cout << "Your currently connected controller has no wave generators." << std::endl;
        }

        PI_CloseConnection (iD);
        std::cout << "Closing connection." << std::endl;
    }
    catch (std::runtime_error e)
    {
        CloseConnectionWithComment (iD, e.what ());
        return 1;
    }

    return 0;
}

double GetMinPositionLimit (int ID, std::string axis)
{
    double minPosLimit = 0;

    if (!PI_qTMN (ID, axis.c_str (), &minPosLimit))
    {
        throw std::runtime_error ("Unable to get minimum position limit.");
    }

    return minPosLimit;
}

double GetMaxPositionLimit (int ID, std::string axis)
{
    double maxPosLimit;

    if (!PI_qTMX (ID, axis.c_str (), &maxPosLimit))
    {
        throw std::runtime_error ("Unable to get maximum position limit.");
    }

    return maxPosLimit;
}

double CalculateAmplitude (int ID, std::string axis)
{
    double amplitude = (std::abs (GetMinPositionLimit (ID, axis)) + std::abs (GetMaxPositionLimit (ID, axis))) / 4;
    return amplitude;
}

void StartWaveGenerator (int ID, std::string axis, waveGeneratorSettings config)
{
    double amplitudeOfWave = CalculateAmplitude (ID, axis);
    std::cout << "Setting up wave generator: Frequency: " << config.frequency << " Hz, Amplitude: " << amplitudeOfWave << " mm" << std::endl;

    // define waveform

    // query servo cycle time
    double servoCycleTime[1];           // servo cycle time of controller
    unsigned int parametersToQuery[1];
    parametersToQuery[0] = PARAM_ServoUpdateTime;

    if (!PI_qSPA (ID, axis.c_str (), parametersToQuery, servoCycleTime, NULL, 0))
    {
        throw std::runtime_error ("Unable to retrieve servo update time.");
    }

    // The available number of points depends on the controller and the settings of the wave tables.
    // calculate number of point in wavetable from given frequency
    int numberOfPoints = (int)(1 / (servoCycleTime[0] * config.frequency));

    // wavetable contains one segment
    int segmentLength = numberOfPoints;

    // curve center point is the middle of the segment
    int centerPointOfWave = numberOfPoints / 2;

    // send wave table to controller
    if (!PI_WAV_SIN_P (ID,
                       config.waveTableId,
                       config.offsetOfFirstPointInWaveTable,
                       numberOfPoints,
                       config.appendWave,
                       centerPointOfWave,
                       amplitudeOfWave,
                       config.offsetOfWave,
                       segmentLength))
    {
        throw std::runtime_error ("Unable to set up waveform.");
    }

    // link wave table to wave generator
    if (!PI_WSL (ID, &config.id, &config.waveTableId, 1))
    {
        throw std::runtime_error ("Unable to link wave table to wave generator.");
    }

    // set wave table rate
    if (!PI_WTR(ID, &SELECT_ALL_GENERATORS, &config.tableRate, &config.interpolationType, 1))
    {
        throw std::runtime_error ("Unable to set wave table rate.");
    }

    // set wave generator output cycles
    if (!PI_WGC (ID, &config.id, &config.numberOfCycles, 1))
    {
        throw std::runtime_error ("Unable to set wave generator output cycles.");
    }

    std::cout << "Starting wave generator..." << std::endl;

    // move stage to offset of wave prior to starting wave generator
    if (!PI_MOV (ID, axis.c_str (), &config.offsetOfWave))
    {
        throw std::runtime_error ("Unable to move axis to start position.");
    }

    WaitForMotionDone (ID, axis);

    // restart recording as soon as wave generator output starts running
    if (!PI_WGR (ID))
    {
        throw std::runtime_error ("Failed to restart recording.");
    }

    // start wave generator output, data recorder starts simultaneously
    if (!PI_WGO (ID, &config.id, &config.startMode, 1))
    {
        throw std::runtime_error ("Unable to start wave generator output.");
    }

    // initialize provisional array for WG Id, 
    // since PI_IsGeneratorRunning expects an array as parameter
    int isWGRunning[1] = { 1 };                   // status of wave generator (0 = not running, 1 = running)
    int waveGeneratorIdArray[1] = { config.id };

    // query wave generator status
    while (isWGRunning[0] == 1)
    {
        Sleep (100);

        if (!PI_IsGeneratorRunning (ID, waveGeneratorIdArray, isWGRunning, 1))
        {
            throw std::runtime_error ("Unable to retrieve wave generator status.");
        }
    }
}

int OpenConnection (ConnectionType type, std::string hostName)
{
    int iD = 0;
    std::string connType;

    switch (type)
    {
        case Dialog:
            iD = PI_InterfaceSetupDlg ("");
            connType = "Dialog";
            break;
        case RS232:
            iD = PI_ConnectRS232 (1, 115200);
            connType = "RS232";
            break;
        case TCPIP:
            iD = PI_ConnectTCPIP (hostName.c_str (), 50000);
            connType = "TCPIP";
            break;
        case USB:
            char connectedUsbController[1024];
            int noDevices = PI_EnumerateUSB (connectedUsbController, 1024, "");
            std::cout << "Found " << noDevices << " controllers connected via USB, connecting to the first one:" << std::endl << connectedUsbController << std::endl;
            char* pos = strchr (connectedUsbController, '\n');

            if (pos)
            {
                *pos = '\0';
            }

            iD = PI_ConnectUSB (connectedUsbController);
            connType = "USB";
            break;
    }

    if (iD < 0)
    {
        ReportError (iD);
        std::cout << "Connect " << connType << ": ERROR" << std::endl;
    }

    return iD;
}

void PrintControllerIdentification (int iD)
{
    char szIDN[200];

    if (!PI_qIDN (iD, szIDN, 199))
    {
        std::runtime_error ("qIDN failed. Exiting.");
    }
    std::cout << "qIDN returned: " << szIDN << std::endl;
}

void CloseConnectionWithComment (int iD, std::string comment)
{
    std::cout << comment << std::endl;

    ReportError (iD);
    PI_CloseConnection (iD);
}

bool WaitForMotionDone (int ID, std::string axis)
{
    BOOL isMoving = TRUE;

    while (isMoving)
    {
        if (!PI_IsMoving (ID, axis.c_str (), &isMoving))
        {
            throw std::runtime_error ("Unable to retrieve motion state.");
        }

        Sleep (100);
    }

    return true;
}

void SetServoState (int ID, std::string axis, BOOL state)
{
    if (!PI_SVO (ID, axis.c_str (), &state))
    {
        throw std::runtime_error ("SVO failed. Exiting");
    }

    std::cout << "Turned servo " << (state ? "on" : "off") << std::endl;
}

void ReportError (int iD)
{
    int err = PI_GetError (iD);
    char szErrMsg[300];

    if (PI_TranslateError (err, szErrMsg, 299))
    {
        std::cout << "Error " << err << " occurred: " << szErrMsg << std::endl;
    }
}
